pkg/netlink/annotatedb.go

```diff
@@ -0,0 +1,431 @@
+package netlink
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"net"
+	"path"
+	"reflect"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/opiproject/opi-evpn-bridge/pkg/infradb"
+	vn "github.com/vishvananda/netlink"
+)
+
+// annotateDBEntries annonates the database entries
+func annotateDBEntries() {
+	for _, nexthop := range latestNexthop {
+		nexthop = nexthop.annotate()
+		latestNexthop[nexthop.Key] = nexthop
+	}
+	for _, r := range latestRoutes {
+		r = r.annotate()
+		latestRoutes[r.Key] = r
+	}
+
+	for _, m := range latestFDB {
+		m = m.annotate()
+		latestFDB[m.Key] = m
+	}
+	for _, l2n := range latestL2Nexthop {
+		l2n = l2n.annotate()
+		latestL2Nexthop[l2n.Key] = l2n
+	}
+}
+
+// nolint
+func (l2n *L2NexthopStruct) annotate() *L2NexthopStruct {
+	// Annotate certain L2 Nexthops with additional information from LB and GRD
+	l2n.Metadata = make(map[interface{}]interface{})
+	LB := l2n.lb
+	if LB != nil {
+		if l2n.Type == SVI {
+			l2n.Metadata["vrf_id"] = *LB.Spec.Vni
+		} else if l2n.Type == VXLAN {
+			//# Remote EVPN MAC address learned on the VXLAN interface
+			//# The L2 nexthop must have a destination IP address in dst
+			l2n.Resolved = false
+			l2n.Metadata["local_vtep_ip"] = *LB.Spec.VtepIP
+			l2n.Metadata["remote_vtep_ip"] = l2n.Dst
+			l2n.Metadata["vni"] = *LB.Spec.Vni
+			//# The below physical nexthops are needed to transmit the VXLAN-encapsuleted packets
+			//# directly from the nexthop table to a physical port (and avoid another recirculation
+			//# for route lookup in the GRD table.)
+			VRF, _ := infradb.GetVrf("//network.opiproject.org/vrfs/GRD")
+			r, ok := lookupRoute(l2n.Dst, VRF)
+			if ok {
+				//  # For now pick the first physical nexthop (no ECMP yet)
+				phyNh := r.Nexthops[0]
+				link, _ := vn.LinkByName(NameIndex[phyNh.nexthop.LinkIndex])
+				l2n.Metadata["phy_smac"] = link.Attrs().HardwareAddr.String()
+				l2n.Metadata["egress_vport"] = phyPorts[NameIndex[phyNh.nexthop.LinkIndex]]
+				if phyNh.Neighbor != nil {
+					if phyNh.Neighbor.Type == PHY {
+						l2n.Metadata["phy_dmac"] = phyNh.Neighbor.Neigh0.HardwareAddr.String()
+						l2n.Resolved = true
+					} else {
+						log.Printf("netlink: Error: Neighbor type not PHY\n")
+					}
+				}
+			}
+		} else if l2n.Type == BRIDGEPORT {
+			// BridgePort as L2 nexthop
+			l2n.Metadata["vport_id"] = l2n.bp.Metadata.VPort
+			l2n.Metadata["portType"] = l2n.bp.Spec.Ptype
+		}
+	}
+	return l2n
+}
+
+// annotate the route
+func (fdb *FdbEntryStruct) annotate() *FdbEntryStruct {
```

Comment:
Done

Reply:
Not done.
I have a mistake in the original comment specifying a wrong receiver type. The code above is updated.

What I want to see is a more object-oriented(in Go sense) approach to remove all those ifs with type checks here and there and have more straightforward logic.
And I want to see more encapsulation where an object can access only members kept within this object. And no(ideally) or reduced number of global variables. Global variables What we have now is very close to [object orgy](https://en.wikipedia.org/wiki/Object_orgy) where in any function literally anything can happen, and it can access any other variable inside. All those endless map manipulations all around the code are very hard to comprehend